### PR TITLE
Fix issue 482

### PR DESCRIPTION
**Pull request by hra687261**
Purification of functional array expressions, fixes issue #482.

### CODE REVIEW

File: src/lib/structures/expr.ml [hunk 1]
```diff
@@ -2433,6 +2433,30 @@ module Purification = struct
           | _, (B_lemma _ | B_skolem _ | B_none | B_let _) -> assert false
         end
 
+      (* When e is an access to a functional array
+         in which the stored values are booleans *)
+      | Sy.Op Get ->
+        begin match e.xs with
+          | [x; y] ->
```
Comment (Gbury) [inline src/lib/structures/expr.ml]: Please give more informative names to these variables. For instance, to better distinguish which is the expression for the array, and which is the expression for the looked-up index.

File: src/lib/structures/expr.ml [hunk 2]
```diff
@@ -2433,6 +2433,30 @@ module Purification = struct
...
+            let x' =
+              if is_pure x then x
+              else
+                let x', lets = purify_term x SMap.empty in
+                mk_lifted x' lets
```
Comment (Gbury) [inline src/lib/structures/expr.ml]: wouldn't it be better to lift the required let-bindings at the top ? i.e. something like: `let x', x'_lets = .... in let y', y'_lets = ... in mk_lifted ...` ?

**Review by Gbury — CHANGES_REQUESTED** [review]
I tihk the fix is good, but I left a few comments. Also, while we're at it, could you change the `assert false` that are present in the purification functions into e.g. `failwith "unexpected expression in purification function"` so that the error message is less cryptic when such bugs occur again ?

File: src/lib/structures/expr.ml (hunk 1 above)
Comment (hra687261) [reply src/lib/structures/expr.ml]: Ok, I replaced ```x``` with ```fa``` for functional array, and ```y``` with ```i``` for index.

File: src/lib/structures/expr.ml (hunk 2 above)
Comment (hra687261) [reply src/lib/structures/expr.ml]: Yes, you're right. It's done.

**Review by Gbury — APPROVED** [review]

**Pull request merged**